audio-echo/app/src/main/java/com/google/sample/echo/PermissionRequestFragment.java

```diff
@@ -0,0 +1,167 @@
+            return;
+        }
+
+        theInstance.callbackPtr = callbackPtr;
+
+        int permissionCheck = ContextCompat.checkSelfPermission(theInstance.getActivity(), permission);
+        if (permissionCheck == PackageManager.PERMISSION_DENIED) {
+            Log.d(TAG,"self permission is denied, requesting");
+            // request runtime permission
+            if (ActivityCompat.shouldShowRequestPermissionRationale(theInstance.getActivity(),
+                    permission)) {
+
+                final Activity thisActivity = theInstance.getActivity();
+                View view = thisActivity.getCurrentFocus();
+                if (view == null) {
+                    view = thisActivity.getWindow().getDecorView();
+                }
+                Snackbar.make(view,
+                        rationale,
+                        Snackbar.LENGTH_INDEFINITE).setAction("OK", new View.OnClickListener() {
+                    @Override
+                    public void onClick(View view1) {
+                        theInstance.requestPermissions(new String[]{permission}, RC_REQUEST_PERMISSION);
+                    }
+                }).show();
+            } else {
+                theInstance.requestPermissions(new String[]{permission}, RC_REQUEST_PERMISSION);
+            }
+        } else {
+            long ptr = theInstance.callbackPtr;
+            theInstance.callbackPtr = 0L;
+            Log.d(TAG,permission + " granted!");
+            handlePermissionResult(PackageManager.PERMISSION_GRANTED, ptr);
+        }
+    }
+
+    /**
+     * Callback for the result from requesting permissions. This method
+     * is invoked for every call on {@link #requestPermissions(String[], int)}.
+     * <p>
+     * <strong>Note:</strong> It is possible that the permissions request interaction
+     * with the user is interrupted. In this case you will receive empty permissions
+     * and results arrays which should be treated as a cancellation.
+     * </p>
+     *
+     * @param requestCode  The request code passed in {@link #requestPermissions(String[], int)}.
+     * @param permissions  The requested permissions. Never null.
+     * @param grantResults The grant results for the corresponding permissions
+     *                     which is either {@link PackageManager#PERMISSION_GRANTED}
+     *                     or {@link PackageManager#PERMISSION_DENIED}. Never null.
+     * @see #requestPermissions(String[], int)
+     */
+    @Override
+    public void onRequestPermissionsResult(int requestCode,
+                                           @NonNull String[] permissions, @NonNull int[] grantResults) {
+
+        if (requestCode == RC_REQUEST_PERMISSION) {
+            long ptr = callbackPtr;
+            if (grantResults.length > 0 && grantResults[0] == PackageManager.PERMISSION_GRANTED) {
+                // finish starting
+                Log.d(TAG,permissions[0] + " granted!");
+                callbackPtr = 0L;
```

Comment:
nit:    
this is for locking purpose, I think.  maybe move it out of the if(){} else {} since it is common. Put it right after you made a copy of it on the top for easy code reading